app/assets/javascripts/restricted_notes/RestrictedNotesPageContainer.js

```diff
@@ -0,0 +1,140 @@
+import _ from 'lodash';
+import PropTypes from '../helpers/prop_types.jsx';
+import {merge} from '../helpers/react_helpers.jsx';
+import React from 'react';
+
+/*
+Holds page state, makes API calls to manipulate it.
+*/
+class RestrictedNotesPageContainer extends React.Component {
+
+  constructor(props) {
+    super(props);
+
+    const serializedData = this.props.serializedData;
+
+    this.state = {
+      // context
+      currentEducator: serializedData.currentEducator,
+      // constants
+      educatorsIndex: serializedData.educatorsIndex,
+      eventNoteTypesIndex: serializedData.eventNoteTypesIndex,
+      // data
+      feed: serializedData.feed,
+      student: serializedData.student,
+      // ui
+      // This map holds the state of network requests for various actions.  This allows UI components to branch on this
+      // and show waiting messages or error messages.
+      // The state of a network request is described with null (no requests in-flight),
+      // 'pending' (a request is currently in-flight),
+      // and 'error' or another value if the request failed.
+      // The keys within `request` hold either a single value describing the state of the request, or a map that describes the
+      // state of requests related to a particular object.
+      requests: {
+        saveNote: null,
+        saveService: null,
+        discontinueService: {}
+      }
+    };
+
+    this.onClickSaveNotes = this.onClickSaveNotes.bind(this);
+    this.onSaveNotesDone = this.onSaveNotesDone.bind(this);
+  }
+
+  componentWillMount(props, state) {
+    const Api = window.shared.Api;
```

Comment:
@Jngai What's the thinking behind moving these from the top of the file down into functions? My understanding is that listing them at the top of the file is the standard way to go because it provides a dependency list up top.

Reply:
With those at the top I was getting `API not a constructor error`. I fixed it by pull those `const` down. Maybe there could be a better way to solve it?

Reply:
@Jngai Yes. I think it's worth investigating that error and figuring out what was causing it.

Reply:
@alexsoble just pull them above `class` and page won't render.

Reply:
@Jngai Hmm, why do you think that is?

Reply:
@alexsoble I am not sure yet but `const` is undefined and can't get there somehow.

Reply:
@Jngai Let me know what you find! If you're still stuck after spending an hour on it let me know and I'll dive in.

Reply:
@alexsoble ok great

Reply:
@alexsoble I am not sure. Will you try?  I tried putting the `const Api` in the constructor and used `bind`. The module got imported and under console its says its not a constructor. I was looking at `api.jsx` and it feels like it could be written better. I tried rewriting it into import/export to make it easier to import but I was getting syntax errors.

Reply:
@Jngai Let me take a look!

Reply:
I did some console printing: 

![screen shot 2017-12-08 at 3 30 46 pm](https://user-images.githubusercontent.com/3209501/33785943-d3e9ec22-dc2c-11e7-8b02-c95b6fd84953.png)

What's interesting is that `window.shared.Api` evaluates to `undefined`, even though we can see it clearly as an object attached to `window.shared`:

![screen shot 2017-12-08 at 3 30 53 pm](https://user-images.githubusercontent.com/3209501/33785971-eb2b772a-dc2c-11e7-84b5-c2fd43ae37c0.png)

We could dig into this further and solve it. Another approach: Api.jsx doesn't declare any dependencies, meaning it's one of our lowest-level classes. How about we move lower level classes to import/export first? That way we could `import` Api into RestrictedNotesPageContainer, and sidestep debugging this while moving towards our overall goal.

When you say "I was looking at api.jsx and it feels like it could be written better" — that doesn't help me understand your thinking at all.

It's much more useful if you say what specifically you don't like about how the file is written, or suggest some alternatives, or ask a question about why something is written this way as opposed to that way. Without any specifics or alternatives, the statement comes across as unhelpfully negative, because it doesn't to anything to say how we might write the file in a better way.